teps/0154-concise-remote-resolver-syntax.md

```diff
@@ -0,0 +1,340 @@
+-->
+
+### Flexibility
+
+<!--
+https://github.com/tektoncd/community/blob/main/design-principles.md#flexibility
+
+- Are there dependencies that need to be pulled in for this proposal to work? What support or maintenance would be
+required for these dependencies?
+- Are we coupling two or more Tekton projects in this proposal (e.g. coupling Pipelines to Chains)?
+- Are we coupling Tekton and other projects (e.g. Knative, Sigstore) in this proposal?
+- What is the impact of the coupling to operators e.g. maintenance & end-to-end testing?
+- Are there opinionated choices being made in this proposal? If so, are they necessary and can users extend it with
+their own choices?
+-->
+
+### Conformance
+
+<!--
+https://github.com/tektoncd/community/blob/main/design-principles.md#conformance
+
+- Does this proposal require the user to understand how the Tekton API is implemented?
+- Does this proposal introduce additional Kubernetes concepts into the API? If so, is this necessary?
+- If the API is changing as a result of this proposal, what updates are needed to the
+[API spec](https://github.com/tektoncd/pipeline/blob/main/docs/api-spec.md)?
+-->
+
+### User Experience
+
+<!--
+(optional)
+
+Consideration about the user experience. Depending on the area of change,
+users may be Task and Pipeline editors, they may trigger TaskRuns and
+PipelineRuns or they may be responsible for monitoring the execution of runs,
+via CLI, dashboard or a monitoring system.
+
+Consider including folks that also work on CLI and dashboard.
+-->
+
+### Performance
+
+<!--
+(optional)
+
+Consider which use cases are impacted by this change and what are their
+performance requirements.
+- What impact does this change have on the start-up time and execution time
+of TaskRuns and PipelineRuns?
+- What impact does it have on the resource footprint of Tekton controllers
+as well as TaskRuns and PipelineRuns?
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Tekton ecosystem. Consider including folks that also work outside the WGs
+or subproject.
+- How will security be reviewed and by whom?
+- How will UX be reviewed and by whom?
+-->
+
+### Drawbacks
+
+<!--
+Why should this TEP _not_ be implemented?
+-->
+
+## Alternatives
+
+## Future Work
+
+The current remote resolution requires the resolver name to be provided by the Task/Pipeline author. This means that if the same pipeline/Task is applied on another cluster where the resolver names are different then they would not work. It would be nice to make the `resolver-name` optional and let Tekton figure out how to route through the resolvers on the cluster and pass the `url`. This currently adds significant complexity to remote resolution and may require a more careful design. We therefore think that this will make an excellent future work item.  
```

Comment:
I'll let @wlynch comment but the idea here was that different resolvers subscribe to schemes (e.g. https, git+https, etc.). The Tekton pipelinerun/taskrun controller parses the scheme and then forwards the request to the resolvers that have subscribed to this scheme one at a time based on some form of priority. 

e.g. a `url` of `https://github.com/tektoncd/pipeline` would first be tried by the git-resolver and if that fails, we would send it to the http resolver because both those resolvers have subscribed to the `https` scheme.

Because something like this involves significant change in the existing architecture, we scoped it out of this TEP into future work. 

So today, the request for:
```
taskRef:
  resolver: git
  url: https://github.com/tektoncd/pipeline
```
Would go explicitly to the git resolver. Once we have this future work, the `resolver: git` would be optional and Tekton would try forwarding the requests.

Reply:
I like the idea of "scheme subscription" and priority (we'll have to be very careful) 👼🏼 But I would definitely include it in that proposal. It does require some significant change in the architecture, but we'll have to do it at some point, so let's describe here what we want long-term and then let's figure out how we go there.

Reply:
I like that idea as well big +1 on making it more easy for end users to don't care about the concept of resolvers and having to choose one...

Reply:
Ok. I can expand on it a bit more and try to lay out a plan of how we can get there. I didn't want to go into lots of details since they are still TBD and beyond the scope of this TEP.

Reply:
I would really love to expand the scope of this TEP to include them 👼🏼

Reply:
The reason I put the scheme subscription based solution out of scope for this TEP is because it significantly increases the complexity and the timeline to deliver it. I don't have the capacity to drill into that right now and so I didn't want to expand the scope of I know that I cant deliver on it.

However, I wanted to convince you that the current TEP will get us half way there towards the subscription based approach. 

 The format of the `name/url` (whatever we call it) starts with a scheme. 
 For now, we can't tell based on the scheme as to which resolver to try. That's why we want to pair it with the existing `resolver` field which informs us which resolver to send the `name/url` to. 
 In the future, we can make the `resolver` field optional when we have the scheme subscription based approach implemented. Even then, if a user passes the `resolver` field, then we know exactly which resolver to pass the request to so it is backwards compatible. 
```yaml
 taskRef:
   name: "https://github.com/tektoncd/pipeline@main#examples/v1/taskrun/hello.yaml"
   resolver: git # in the future this is optional.
```

Reply:
However, if others were interested in collaborating with me on this, then its definitely doable.

Reply:
We can also do this in two stages where in stage 1, we do what's currently suggested and stage 2, we implement the scheme based subscription.

Reply:
Resolving this now since its now in the main scope.